### PR TITLE
Initial client/server networking

### LINKED ISSUE #9: Add network library as a dependency

### LINKED ISSUE #11: Initial network communication

### PR DESCRIPTION
**Pull request by caspernorrbin**
Adds client/server networking within `admirals::net`

Uses ASIO as a submodule, so it needs it needs to be downloaded like the rest of the submodules.

The communication is message-based, with messages from `Message.hpp` being sent. The whole implementation uses templates to allow for user-defined message types. This allows messages to be of any type and be of any size.

Each connection from a client or the server is handled by a connection object from `Connection.hpp`. Each network connection has two connection objects associated with it: One on the client, and one on the server. The functionality is broadly the same for both, with some small differences in specific functions.

Message queues are used for messaging. Each Connection object has an outgoing message queue, and all clients and the server all have queues for incoming messages. The clients/server can thus choose when to read these queues to process incoming messages.

Since the implementation builds on templates, the implementation cannot be in source files. To circumvent this, the implementation is placed in `.tpp` files (source files for templates), with them being included at the bottom of the header. I am very open to potential other solutions for this.

Also included are two test files: One server and one client. This implements a small "game" of an 8x8 grid. Clients can press 'Q' in their console to spawn a "ship" at (0,0), press 'W' to move it to a random location, and press 'E' to quit. The game waits for 2 players to connect before automatically starting. If run on the same computer, both clients will read the same input and both will try to send an action to the server.

resolves #9, resolves #11

### CODE REVIEW

**Review by JoXW100 — COMMENTED** [review]
I feel like the lambda expressions in `Connection.tpp` are difficult to read, and we should extract them to functions.

I also feel like the additional complexity of using enums for `Message::id` is not worth it, by removing the use of enums here, most of the network implementation can be done without using generic classes. 

I have created a branch [networking_without_enum](https://github.com/joelsiks/admirals/tree/networking_without_enum)  where the enum is removed and `.tpp` files are changed into `.cpp` files.

File: src/engine/net/Message.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    // Returns the size of the message body in bytes
+    size_t Size() const { return body.size(); }
+
+    // Overload for the << operator for adding data to the message buffer
+    template <typename DataType>
+    friend Message<T> &operator<<(Message<T> &msg, const DataType &data) {
+        // Check that the data type is trivially copyable
+        static_assert(std::is_standard_layout<DataType>::value,
+                      "Data is too complex to be pushed into vector");
```
Comment (joelsiks) [inline src/engine/net/Message.hpp:23]: Smart!

File: src/engine/net/Server.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+protected:
+    // Called when a client connects
+    virtual bool OnClientConnect(std::shared_ptr<Connection<T>> client) {
+        return false;
+    }
+    // Called when a client disconnects
+    virtual void OnClientDisconnect(std::shared_ptr<Connection<T>> client) {}
+    // Called when a message is received
+    virtual void OnMessage(std::shared_ptr<Connection<T>> client,
+                           Message<T> &message) {}
```
Comment (joelsiks) [inline src/engine/net/Server.hpp:47]: These should probably be purely virtual, i.e `virtual void OnClientDisconnect(std::shared_ptr<Connection<T>> client) = 0;`

**Review by joelsiks — COMMENTED** [review]
I also think that the lambda functions are somewhat hard to read. One alternative is to move the lambda function into its own function right above the associated function it is used in. That way the deep indentation isn't a problem any more.

File: src/engine/net/Client.tpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+template <typename T> Client<T>::Client() {}
+
+template <typename T> Client<T>::~Client() { Disconnect(); }
+
+template <typename T>
+bool Client<T>::Connect(const std::string &host, const std::string &port) {
+    try {
+        asio::ip::tcp::resolver resolver(m_io_context);
+        asio::ip::tcp::resolver::results_type endpoints =
+            resolver.resolve(host, port);
```
[the remark this answers is not in the document]
Comment (caspernorrbin) [reply src/engine/net/Client.tpp]: `resolve()` returns a list of 0 or more endpoints, and can potentially have more than one endpoint. When later connecting, one endpoint is used and referenced.

File: src/engine/net/Client.hpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    void Send(const Message<T> &message);
+
+    // Returns the incoming message queue
+    MessageQueue<OwnedMessage<T>> &Incoming();
+
+private:
+    asio::io_context m_io_context;
+    std::thread m_context_thread;
+    std::unique_ptr<Connection<T>> m_connection;
+    MessageQueue<OwnedMessage<T>> m_incoming_messages;
```
[the remark this answers is not in the document]
Comment (caspernorrbin) [reply src/engine/net/Client.hpp]: Fixed in a839b8f

File: src/engine/net/Connection.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    void ReadHeader();
+    void ReadBody();
+    void AddToIncomingMessageQueue();
+    void WriteHeader();
+    void WriteBody();
+    uint64_t HashValidation(uint64_t hash);
+    void WriteValidation();
+    void ReadValidation(admirals::net::Server<T> *server = nullptr);
+
+private:
```
[the remark this answers is not in the document]
Comment (caspernorrbin) [reply src/engine/net/Connection.hpp:66]: It is a remnant of a previous implementation, removed in 13a5840

File: src/engine/net/Connection.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    Owner m_owner_type = Owner::SERVER;
+    uint32_t m_id = 0;
+
+    uint64_t m_validation_out = 0;
+    uint64_t m_validation_in = 0;
+
+    // Used when validating a connection
+    Server<T> *m_server = nullptr;
+    void (Server<T>::*m_validation_function)(std::shared_ptr<Connection<T>>) =
```
[the remark this answers is not in the document]
Comment (caspernorrbin) [reply src/engine/net/Connection.hpp]: Changed to a `std::function` in 13a5840

File: src/engine/net/Connection.tpp [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+namespace net {
+
+template <typename T>
+Connection<T>::Connection(
+    Owner parent, asio::io_context &io_context, asio::ip::tcp::socket socket,
+    MessageQueue<OwnedMessage<T>> &incoming_messages, Server<T> *server,
+    void (Server<T>::*validation_function)(std::shared_ptr<Connection<T>>))
+    : m_io_context(io_context), m_socket(std::move(socket)),
+      m_incoming_messages(incoming_messages), m_owner_type(parent),
+      m_server(server), m_validation_function(validation_function) {
```
[the remark this answers is not in the document]
Comment (caspernorrbin) [reply src/engine/net/Connection.tpp]: Moved most to the initializations to the constructor body instead in 13a5840

File: src/engine/net/Server.hpp (hunk 2 above)
Comment (caspernorrbin) [reply src/engine/net/Server.hpp]: Fixed in 13a5840

**Comment by caspernorrbin** [comment]
Rebased the [networking_without_enum](https://github.com/joelsiks/admirals/tree/networking_without_enum) branch onto this one to remove templates as done by @JoXW100 .

The read/write lambda functions have been separated and moved directly above the respective functions. I am unsure of where to place these, and am open to suggestions. Additionally, the smaller fixes have been added to further increase readability.

File: src/engine/net/Connection.cpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+        m_socket.close();
+    }
+}
+
+// Reads the full body of the message
+void Connection::ReadBody() {
+    asio::async_read(
+        m_socket,
+        asio::buffer(m_messageBuffer.body.data(), m_messageBuffer.body.size()),
+        [this](std::error_code ec, std::size_t length) { HandleReadBody(ec); });
```
Comment (JoXW100) [inline src/engine/net/Connection.cpp:120]: Here aswell...

File: src/engine/net/Connection.cpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+// Writes the header of the first message in the outgoing queue
+void Connection::WriteHeader() {
+    asio::async_write(
+        m_socket,
+        asio::buffer(&m_outgoingMessages.Front().header, sizeof(MessageHeader)),
+        [this](std::error_code ec, std::size_t length) {
+            HandleWriteHeader(ec);
+        });
```
Comment (JoXW100) [inline src/engine/net/Connection.cpp:165]: Here aswell...

File: src/engine/net/Connection.cpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+// Writes the body of the first message in the outgoing queue
+void Connection::WriteBody() {
+    asio::async_write(m_socket,
+                      asio::buffer(m_outgoingMessages.Front().body.data(),
+                                   m_outgoingMessages.Front().body.size()),
+                      [this](std::error_code ec, std::size_t length) {
+                          HandleWriteBody(ec);
+                      });
```
Comment (JoXW100) [inline src/engine/net/Connection.cpp:191]: Here aswell...

File: src/engine/net/Connection.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    }
+}
+
+// Writes the (either generated or received) validation number
+void Connection::WriteValidation() {
+    asio::async_write(m_socket,
+                      asio::buffer(&m_validationOut, sizeof(uint64_t)),
+                      [this](std::error_code ec, std::size_t length) {
+                          HandleWriteValidation(ec);
+                      });
```
Comment (JoXW100) [inline src/engine/net/Connection.cpp:226]: Here aswell...

File: src/engine/net/Connection.cpp [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    }
+}
+
+// Reads the (either hashed or generated) validation number and validates it (if
+// the connection is a server)
+void Connection::ReadValidation() {
+    asio::async_read(m_socket, asio::buffer(&m_validationIn, sizeof(uint64_t)),
+                     [this](std::error_code ec, std::size_t length) {
+                         HandleReadValidation(ec);
+                     });
```
Comment (JoXW100) [inline src/engine/net/Connection.cpp:265]: Here aswell...

File: src/engine/net/Connection.hpp [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    Owner m_ownerType = Owner::SERVER;
+    uint32_t m_id = 0;
+
+    uint64_t m_validationOut = 0;
+    uint64_t m_validationIn = 0;
+
+    // Used when validating a connection
+    Server *m_server = nullptr;
+    // void (Server::*m_validationFunction)(std::shared_ptr<Connection>) =
+    // nullptr;
```
Comment (JoXW100) [inline src/engine/net/Connection.hpp:83]: Remove this comment

File: src/engine/net/Server.cpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+                  << std::endl;
+
+        // Create a new connection object for the client
+        std::shared_ptr<Connection> new_connection =
+            std::make_shared<Connection>(
+                Connection::Owner::SERVER, m_ioContext, std::move(socket),
+                m_incomingMessages, this,
+                [this](std::shared_ptr<Connection> connection) {
+                    this->OnClientValidated(connection);
+                });
```
Comment (JoXW100) [inline src/engine/net/Server.cpp:46]: Here aswell...

File: src/engine/net/Server.cpp [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        std::cout << "Connection error: " << ec.message() << std::endl;
+    }
+    WaitForClientConnection();
+}
+
+void Server::WaitForClientConnection() {
+    m_acceptor.async_accept(
+        [this](std::error_code ec, asio::ip::tcp::socket socket) {
+            HandleAcceptedConnection(ec, std::move(socket));
+        });
```
Comment (JoXW100) [inline src/engine/net/Server.cpp:67]: Here aswell...

File: src/engine/net/Connection.cpp [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+        m_socket, endpoints,
+        [this](std::error_code ec, asio::ip::tcp::endpoint endpoint) {
+            if (!ec) {
+                // Read the validation number from the server and send it back
+                // hashed
+                ReadValidation();
+            } else {
+                std::cout << "Connect failed: " << ec.message() << std::endl;
+            }
+        });
```
Comment (JoXW100) [inline src/engine/net/Connection.cpp:48]: I think we should extract this lambda expression. This one is not that badly formatted, but i think we should stay away from lambda expressions with multiline bodies.

**Review by JoXW100 — CHANGES_REQUESTED** [review]
Looks good, but we should remove most of the leftover lambda expressions. Did you figure out the cause of the warnings?

I have implemented this in branch `networking_bound_handlers` that can be merged into this if you agree with these changes. The changes can be viewed in Draft #40.

File: src/engine/net/Connection.cpp (hunk 16 above)
Comment (caspernorrbin) [reply src/engine/net/Connection.cpp]: Changed in 713165b

File: src/engine/net/Connection.hpp (hunk 13 above)
Comment (caspernorrbin) [reply src/engine/net/Connection.hpp]: Fixed in 713165b

**Review by JoXW100 — APPROVED** [review]

**Pull request merged**